### PR TITLE
Replace easy Substrings with AsSpan/Slices

### PR DESCRIPTION
**Pull request by stephentoub**
I went through all Substring calls in coreclr and replaced the ones that were straightforward.

cc: @jkotas, @tarekgh, @danmosemsft

### CODE REVIEW

File: src/mscorlib/shared/System/Globalization/TextInfo.cs [hunk 1]
```diff
@@ -809,11 +809,21 @@ private int AddTitlecaseLetter(ref StringBuilder result, ref string input, int i
...
 
-            // for surrogate pairs do a simple ToUpper operation on the substring
+            // for surrogate pairs do a ToUpper operation on the substring
             if (charLen == 2)
             {
                 // Surrogate pair
-                result.Append(ToUpper(input.Substring(inputIndex, charLen)));
+                ReadOnlySpan<char> src = input.AsSpan(inputIndex, 2);
+                Span<char> dst = stackalloc char[2];
+                if (GlobalizationMode.Invariant)
```
Comment (tarekgh) [inline src/mscorlib/shared/System/Globalization/TextInfo.cs]: >GlobalizationMode.Invariant [](start = 20, length = 27)

use _invariantMode instead

File: src/mscorlib/shared/System/Globalization/TextInfo.cs [hunk 2]
```diff
@@ -809,11 +809,21 @@ private int AddTitlecaseLetter(ref StringBuilder result, ref string input, int i
...
+                {
+                    ToUpperAsciiInvariant(src, dst);
```
Comment (tarekgh) [inline src/mscorlib/shared/System/Globalization/TextInfo.cs]: >ToUpperAsciiInvariant(src, dst); [](start = 20, length = 32)

I don't think we need to do that as we are in invariant mode and we know we are handling Surrogate. we can just append the original characters as these characters will be outside the ascii range anyway,

File: src/mscorlib/shared/System/Reflection/AssemblyNameFormatter.cs [hunk 3]
```diff
@@ -111,8 +111,8 @@ private static void AppendQuoted(this StringBuilder sb, string s)
                         continue;
                     if ((s.Length - i) < escapeReplacement.Length)
                         continue;
-                    string prefix = s.Substring(i, escapeReplacement.Length);
-                    if (prefix == escapeReplacement)
+                    ReadOnlySpan<char> prefix = s.AsSpan(i, escapeReplacement.Length);
+                    if (prefix.Equals(escapeReplacement, StringComparison.Ordinal))
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/mscorlib/shared/System/Reflection/AssemblyNameFormatter.cs]: Or call `prefix.SequenceEquals(...`. It is the lowest level / fastest API.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/mscorlib/shared/System/Reflection/AssemblyNameFormatter.cs]: I'll change it to use SequenceEquals. Thanks.

File: src/mscorlib/shared/System/Globalization/TextInfo.cs (hunk 1 above)
Comment (stephentoub) [reply src/mscorlib/shared/System/Globalization/TextInfo.cs]: Ok

File: src/mscorlib/shared/System/Globalization/TextInfo.cs (hunk 2 above)
Comment (stephentoub) [reply src/mscorlib/shared/System/Globalization/TextInfo.cs]: So you mean change this to something like:
```C#
if (_invariantMode)
{
    result.Append(src);
}
else
{
    Span<char> dst = stackalloc char[2];
    ChangeCase(src, dst, toUpper: true);
    result.Append(dst);
}
```
?

Comment (tarekgh) [reply src/mscorlib/shared/System/Globalization/TextInfo.cs]: Yes.

**Pull request merged**